packages/smooth_app/lib/query/products_preload_helper.dart

```diff
@@ -1,23 +1,26 @@
 import 'dart:io';
 import 'package:openfoodfacts/openfoodfacts.dart';
 import 'package:smooth_app/database/dao_product.dart';
-import 'package:smooth_app/database/local_database.dart';
 import 'package:smooth_app/query/product_query.dart';
 
 class PreloadDataHelper {
-  PreloadDataHelper(this.localDatabase);
-  LocalDatabase localDatabase;
-
-  Future<String> preloadData() async {
-    final DaoProduct daoProduct = DaoProduct(localDatabase);
+  PreloadDataHelper(this.daoProduct);
+  DaoProduct daoProduct;
+  Future<String> getTopProducts() async {
+    List<String> allProductCodes = await daoProduct.getAllKeys();
+    final Map<String, Product> allProducts =
+        await daoProduct.getAll(allProductCodes);
```

Comment:
That means you're loading ALL the database.
I know it's only for dev mode, but that's not something I would recommend.
What about getting all the barcodes instead of barcodes AND products.
For the level of granularity we want that's good enough I think.

Reply:
@monsieurtanuki 
Just to clarify, this is not always supposed to be in dev mode, after further fine-tuning and doing this task in background, we are probably thinking of showing a menu in onboarding screens to let preload data, but not just 100o, rather in numbers of 10k.

And why I loaded the products as well is because there is no way where I can be sure that which barcodes will have knowledge panels or not. I tried thinking of alternate solutions even using `daoproductlist`, would be helpful for me to get an alternate idea to do the same without loading all the products but with just the barcode strings

Reply:
I guess a SQL query for getting all the barcode with `Products having knowledge panel` could have probably helped here.
But since we already don't do much of magic with SQL, can't say what else we can do here

Reply:
Okay I get all the keys now I updated the database with top 1000 products
Now, if I do again try to get the top 1000 products, just for sense of updating it,
When I do `DaoProduct.getAllKeys()`, I will get the list of barcodes along with the top 1000 products, so now the list of barcodes will have more than 1000, and hence I can't update it. Ryt?
I hope now you can understand the problem...

Reply:
No I don't understand the problem.

What I suggest is OK if you want:
* to get the top 1000 products, 
* to avoid to "erase" the KP of existing products

What else do you want? To refresh the top 1000 products? To go beyond the first 1000 products?

Reply:
Oops, I skipped that one:

> Just to clarify, this is not always supposed to be in dev mode, after further fine-tuning and doing this task in background, we are probably thinking of showing a menu in onboarding screens to let preload data, but not just 100o, rather in numbers of 10k.

But now we're on dev mode, so let's focus on that. Background task later.
Regarding putting that in onboarding, that's a bold move considering the current popularity of the onboarding among the users.
Regarding 10K bulk load, I don't see how you can skip the "download first, then refresh locally if relevant, then the next 10K" for the whole server data.

Reply:
@monsieurtanuki 
About your query to not load all products to check if they have knowledge panels or not
What I propose is we add another column in the table with name `"isHalfProduct"`, and we give it a default value of false 
And when pushing a product to db, we can say that it is not fully downloaded by just passing an argument as true.

Would be best if you could suggest me if I should go ahead with this approach??

Reply:
> What I propose is we add another column in the table with name `"isHalfProduct"`, and we give it a default value of false

I don't think it's a good idea to add computed columns to `sqflite`. In general that's a pain in the neck to add columns: not just adding them, but then initializing them, and keep the init script running forever, and then you add another column and you wind up with several scripts that run forever.
Especially in this case: there's not such thing as `isHalfProduct`. Unless that means "null knowledge panels" on August 5th, 2022. But what next?

My suggestion: you _only_ have to code a single method that will solve _all_ your problems. And luckily, that's easy to code, and to test as a background method:
* 2 parameters, `DaoProduct` and `Product newProduct`
* you try to get from `DaoProduct` a `Product` with `barcode = newProduct.barcode!`
* if there's indeed already one product, you check its `knowledgePanels` field. If it's not null, you return `false` (that's data that you don't want to overwrite)
* else you upsert your `newProduct` and return `true` (that's data that either did not exist locally, or that didn't include kwowledge panels so your new version is better)

Et voilà.

Reply:
Done that, plz re review

Reply:
ping @monsieurtanuki